vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,787 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/reductions/eigen_memory_tree.h"
+
+#include "vw/common/future_compat.h"
+#include "vw/config/options.h"
+#include "vw/core/array_parameters.h"
+#include "vw/core/example.h"
+#include "vw/core/learner.h"
+#include "vw/core/numeric_casts.h"
+#include "vw/core/rand48.h"
+#include "vw/core/rand_state.h"
+#include "vw/core/setup_base.h"
+#include "vw/core/v_array.h"
+#include "vw/core/vw.h"
+#include "vw/io/logger.h"
+
+#include <algorithm>
+#include <cfloat>
+#include <climits>
+#include <cmath>
+#include <cstdio>
+#include <ctime>
+#include <list>
+#include <memory>
+#include <sstream>
+#include <type_traits>
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+
+namespace VW
+{
+namespace reductions
+{
+namespace eigen_memory_tree
+{
+////////////////////////////definitions/////////////////////
+////////////////////////////////////////////////////////////
+emt_example::emt_example() { label = 0; }
+
+emt_example::emt_example(VW::workspace& all, VW::example* ex)
+{
+  label = ex->l.multi.label;
+
+  std::vector<std::vector<VW::namespace_index>>* full_interactions = ex->interactions;
+  std::vector<std::vector<VW::namespace_index>> base_interactions;
+
+  ex->interactions = &base_interactions;
+  auto* ex1 = VW::flatten_sort_example(all, ex);
+  for (auto& f : ex1->fs) { base.emplace_back(f.index(), f.value()); }
+  VW::free_flatten_example(ex1);
+
+  ex->interactions = full_interactions;
+  auto* ex2 = VW::flatten_sort_example(all, ex);
+  for (auto& f : ex2->fs) { full.emplace_back(f.index(), f.value()); }
+  VW::free_flatten_example(ex2);
+}
+
+emt_lru::emt_lru(unsigned long max_size) { (*this).max_size = max_size; }
+
+emt_lru::K emt_lru::bound(emt_lru::K item)
+{
+  if (max_size == 0) { return nullptr; }
+
+  auto item_map_reference = map.find(item);
+
+  // item is not in map so we add it to our list
+  if (item_map_reference == map.end())
+  {
+    list.push_front(item);
+    map.emplace(std::pair<K, V>(item, list.begin()));
+  }
+  // item is in map already so we move it to the front of the line
+  else
+  {
+    V item_list_reference = (*item_map_reference).second;
+    if (item_list_reference != list.begin())
+    {
+      list.splice(list.begin(), list, item_list_reference, std::next(item_list_reference));
+    }
+  }
+
+  if (list.size() > max_size)
+  {
+    K last_value = list.back();
+    list.pop_back();
+    map.erase(last_value);
+    return last_value;
+  }
+  else { return nullptr; }
+}
+
+emt_node::emt_node()
+{
+  left = nullptr;
+  right = nullptr;
+  router_decision = 0;
+}
+
+emt_tree::emt_tree()
+{
+  leaf_split = 100;
+
+  scorer_type = emt_scorer_type::self_consistent_rank;
+  router_type = emt_router_type::eigen;
+
+  begin = 0;
+
+  all = nullptr;
+  ex = nullptr;
```

Comment:
can be removed?

Reply:
Or you mean begin? It looks like it.

Reply:
If you see any small refactors like this feel free to make changes yourself.

Reply:
I am sorry for short comment. I was meaning "ex=nullptr" which is initialized 2 lines later.